### PR TITLE
#1246 Add autocompletion on gem search inputs

### PR DESCRIPTION
**Pull request by clamoris**
Issue: https://github.com/rubygems/rubygems.org/issues/1246

Added autocompletion for the home search bar using Elasticsearch and typeahead.js.

photo_2016-06-08_12-01-22

### CODE REVIEW

File: app/controllers/searches_controller.rb [hunk 1]
```diff
@@ -13,6 +13,12 @@ def show
     redirect_to rubygem_path(@exact_match) if @gems == [@exact_match]
   end
 
+  def autocomplete
+    render json: Rubygem.search(params[:query], es: true).map(&:name)
```
Comment (clamoris) [inline app/controllers/searches_controller.rb]: Here I retrieve up to 30 records but show only 5 on UI. On my machine, there is no difference between 5 and 30 records so I leave the defaults.

File: app/models/concerns/rubygem_searchable.rb [hunk 2]
```diff
@@ -32,17 +32,31 @@ def as_indexed_json(_options = {})
...
+                   tokenizer: 'special_characters',
+                   filter: %w(lowercase autocomplete_ngram)
+                 }
                }
              }
 
     mapping do
       indexes :name, type: 'multi_field' do
-        indexes :name, analyzer: 'rubygem'
+        indexes :name, index_analyzer: 'autocomplete_index', search_analyzer: 'whitespace'
```
Comment (sonalkr132) [inline app/models/concerns/rubygem_searchable.rb]: `index_analyzer` is [removed in ES 2.0
](https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking_20_mapping_changes.html#_analyzer_mappings)

I understand that contribution.md does not specify ES version, but I would advocate that we use the latest ES 2.3. ES 2 has so many improvements.

**Comment by sonalkr132** [comment]
>  If it suits you, I'll extend header search with the same autocompletion.

You will have to wait for a response from rubygems team for approval.
I think a screenshot of what it looks like would be nice.

File: app/models/concerns/rubygem_searchable.rb (hunk 2 above)
Comment (clamoris) [inline app/models/concerns/rubygem_searchable.rb]: Actually CONTRIBUTING.md DO specify version

> Install Elastic Search 1.5.2: https://www.elastic.co/downloads/past-releases/elasticsearch-1-5-2

And just `analyzer` wont't work on 1.5. Is the guideline outdated?

Comment (sonalkr132) [inline app/models/concerns/rubygem_searchable.rb]: > Is the guideline outdated?

No. Turns our we are stuck with ES 1.5 because of aws :angry:

**Comment by clamoris** [comment]
@homu @sonalkr132 @dwradcliffe 

Rebased on master, typeahead added to header search as well.

**Comment by dwradcliffe** [comment]
Thanks! We will review this soon. There are a couple other infrastructure things at the top of the list before we ship these search enhancements.

**Comment by clamoris** [comment]
@dwradcliffe ping =)

Maybe I could help with "other infrastructure things" as well?

**Comment by matiaskorhonen** [comment]
Not sure if this is out of scope for this PR, but hopefully the suggestions will also be added to the OpenSearch description (https://developer.mozilla.org/en-US/docs/Supporting_search_suggestions_in_search_plugins) once the ElasticSearch stuff gets the go-ahead from the rubygems team…

File: Gemfile [hunk 3]
```diff
@@ -45,6 +45,7 @@ gem 'elasticsearch-rails', '~> 0.1.7'
 gem 'elasticsearch-dsl', '~> 0.1.2'
 gem 'xml-simple'
 gem 'yajl-ruby', require: 'yajl'
+gem 'twitter-typeahead-rails'
```
Comment (jvanbaarsen) [inline Gemfile]: Do we want to lock down a version here?

File: app/assets/javascripts/search_typeahead.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    });
+
+    $('#home_query, .header__search').typeahead({
+        minLength: 2,
+        highlight: true
+    }, {
+        name: "home_query",
+        source: suggester
+    });
+});
```
Comment (jvanbaarsen) [inline app/assets/javascripts/search_typeahead.js]: Not sure if this is a problem with rubygems codestyle, but there is a missing newline at the end of the file

File: app/assets/stylesheets/modules/home.css [hunk 5]
```diff
@@ -49,6 +49,7 @@
   margin-left: auto;
   position: relative;
   width: 90%;
+  margin-top: 30px;
```
Comment (jvanbaarsen) [inline app/assets/stylesheets/modules/home.css]: What do you think about moving this one a couple of lines to the top (Underneath line 49), that way the `margin`s are grouped together.

File: app/controllers/searches_controller.rb (hunk 1 above)
Comment (jvanbaarsen) [inline app/controllers/searches_controller.rb]: What do you mean with there is no difference?

**Comment by sonalkr132** [comment]
Closing this for lack of response. 

PS: last commit on twitter/typeahead.js was more than 3 years ago. If you (or anyone else) still want to work on this, I would very much prefer it if we did this without adding any dependency.

**Pull request closed without merging**